Clarify how to override default components that use multiple slots.

jwoyo opened the pull request:
The documentation lacked an explanation of how to override default components that use named slots. Passing all slots through is necessary. Omitting this step breaks the layout for example when user conditionally changes PageFrame.astro.

The commit adds an explanation and a code example.

<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

```astro
---
// this could be: src/components/ConditionalPageFrame.astro
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/PageFrame.astro';
const isHomepage = Astro.props.slug === '';
---
{
	isHomepage ? (
		<section>My great landing page 🌟</section>
	) : (
		<Default {...Astro.props}>
			<slot />
			<Fragment slot="header"><slot name="header"/></Fragment>
			<Fragment slot="sidebar"><slot name="sidebar"/></Fragment>
		</Default>
	)
}
```

<!--
Here’s what will happen next:
One or more of our maintainers will take a look and may ask you to make changes.
We try to be responsive, but don’t worry if this takes a day or two.
-->

astrobot-houston commented:
Hello! Thank you for opening your **first PR** to Starlight! ✨

Here’s what will happen next:

1. Our GitHub bots will run to check your changes.
   If they spot any issues you will see some error messages on this PR.
   Don’t hesitate to ask any questions if you’re not sure what these mean!

2. In a few minutes, you’ll be able to see a preview of your changes on Netlify 🤩

3. One or more of our maintainers will take a look and may ask you to make changes.
   We try to be responsive, but don’t worry if this takes a few days.

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| Locale | File                                                                                                                                                                                      | Note                                                       |
| ------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| en     | [guides/overriding-components.mdx](https://github.com/withastro/starlight/blob/93b019826d5fd7e62af3f2ed34386155e8bdd90b/docs%2Fsrc%2Fcontent%2Fdocs%2Fguides%2Foverriding-components.mdx) | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

HiDeoo commented:
Thanks for your contribution :raised_hands:

This is definitely something that we do not explicitly document and has come up a few times already. I'd like to get other people's opinion on this but my initial thought would be that this would better fit in the ["Reuse a built-in component"](https://starlight.astro.build/guides/overriding-components/#reuse-a-built-in-component) section rather than the "Use page data" one. I'm also not sure we would need an entire new code example for this as this would be mostly identical to other ones we already have but we could link to the [relevant Astro Docs section](https://docs.astro.build/en/basics/astro-components/#transferring-slots).

I think my idea would be to add a third bullet point to the ["Reuse a built-in component"](https://starlight.astro.build/guides/overriding-components/#reuse-a-built-in-component) unordered list that could look something like this:

> - If the built-in component you are reusing also contains [nameed slots](https://docs.astro.build/en/basics/astro-components/#named-slots), e.g. like the [`TwoColumnContent`](https://github.com/withastro/starlight/blob/main/packages/starlight/components/TwoColumnContent.astro) component which uses a `right-sidebar` slot, you need to [transfer](https://docs.astro.build/en/basics/astro-components/#transferring-slots) theses as well using the `<slot name="right-sidebar" slot="right-sidebar" />` syntax.

Another possibility that comes to mind is to reference additional named slots in the ["Overrides Reference"](https://deploy-preview-2386--astro-starlight.netlify.app/reference/overrides/) page for the few components that have them. At the moment, only `TwoColumnContent` and `PageFrame` have named slots.

For example, below the "Default component: TwoColumnContent.astro" line, it could say:

> **Default component:** [`TwoColumnContent.astro`](https://deploy-preview-2386--astro-starlight.netlify.app/reference/overrides/#twocolumncontent:~:text=Default%20component%3A%20TwoColumnContent.astro)
> **Named slots:** `right-sidebar`

Like I said, I'd like to get your and other people's opinion on this before I make any changes. Let me know what you think.

jwoyo commented:
Thanks @HiDeoo for reviewing the proposed change. I agree that the information fits better as a third bullet point under [Reuse a built-in component](https://starlight.astro.build/guides/overriding-components/#reuse-a-built-in-component). Also referencing the named slots in the [Overrides Reference](https://deploy-preview-2386--astro-starlight.netlify.app/reference/overrides/#twocolumncontent:~:text=Default%20component%3A%20TwoColumnContent.astro) makes sense to me as most users probably won't check in the overridden file in depth and would appreciate a hint that they need to handle the slots.

I'll wait for others' opinions as well and proceed with your suggested change in this PR if I don't hear otherwise.

jwoyo commented:
Hey @HiDeoo. As I haven't heard any other opinions on that I've made the suggested changes:

- Removed the additional code example
- Added a third bullet point containing the code example with `TwoColumnContent`
- Added the **named slots:** line to the [Override References](https://deploy-preview-2386--astro-starlight.netlify.app/reference/overrides/)

HiDeoo commented on docs/src/content/docs/guides/overriding-components.mdx:
```diff
@@ -81,7 +81,7 @@ When rendering a built-in component inside a custom component:
 - Spread `Astro.props` into it. This makes sure that it receives all the data it needs to render.
 - Add a [`<slot />`](https://docs.astro.build/en/core-concepts/astro-components/#slots) inside the default component. This makes sure that if the component is passed any child elements, Astro knows where to render them.
-
+- If the built-in component you are reusing also contains [named slots](https://docs.astro.build/en/basics/astro-components/#named-slots), you need to transfer them as well. For example, the [`TwoColumnContent`](/reference/overrides/#twocolumncontent) component uses a `right-sidebar` slot. To keep it you need to transfer it using the `<slot name="right-sidebar" slot="right-sidebar" />` syntax (se ...
```
The ideas of the changes that were suggested during the Talking & Doc'ing event were:

- Explicitly mention the only 2 components that have named slots
- Show the example you initially provided with an highlight on the line transferring a slot


````suggestion

If you are reusing the [`PageFrame`](/reference/overrides/#pageframe) or [`TwoColumnContent`](/reference/overrides/#twocolumncontent) components which contain [named slots](https://docs.astro.build/en/basics/astro-components/#named-slots), you also need to [transfer](https://docs.astro.build/en/basics/astro-components/#transferring-slots) these slots as well.

The example below shows a custom component that reuses the `TwoColumnContent` component which contains an additional `right-sidebar` named slot that needs to be transferred:

```astro {9}
---
// src/components/CustomContent.astro
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/components/TwoColumnContent.astro';
---

<Default {...Astro.props}>
	<slot />
	<slot name="right-sidebar" slot="right-sidebar" />
</Default>
```

````

jwoyo commented:
> The general consensus seems to be that both your initial suggestion with a example and some of the changes I suggested are good ideas so I left a new suggestion matching both of them. Let me know what you think about it.

Sounds perfect. Thank you for getting back! I've added the suggested change.

HiDeoo reviewed (approved):
Thanks for kicking this off and your patience. This looks like a great improvement to me with your latest changes :star2:
